In [1]:
import os
import subprocess
import pandas as pd
import glob

In [2]:
pwd


'/Users/gracer/Desktop/fsl_course_data/script'

In [3]:
basepath = '/Users/gracer/Desktop/fsl_course_data'

In [3]:
os.path.exists(os.path.join(basepath, 'fmri1','fluency_task'))

True

In [4]:
for x in glob.glob(os.path.join(basepath, 'fmri1','fluency_task','*.nii.gz')):
    print(x)

/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri.nii.gz
/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/structural_brain.nii.gz
/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri_brain.nii.gz
/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri_brain_mask.nii.gz
/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/structural.nii.gz


raw
-- sub-001> fmri.nii.gz, struct.nii.gz
-- sub-002> fmri.nii.gz, struct.nii.gz

In [19]:
SUBNUM = []
for z in glob.glob(os.path.join(basepath,'fmri1','raw_data','sub-*')):
    y = z.split('/')[-1].split('-')[-1]
    SUBNUM.append(y)

002
001


In [20]:
SUBNUM

['002', '001']

In [21]:
SUBNUM = ['002','001']

In [4]:
cmd = 'fslnvol %s'%subnum


NameError: name 'subnum' is not defined

In [29]:
for sub in SUBNUM:
    SUB = os.path.join(basepath, 'fmri1','raw_data','sub-%s'%sub,'fmri.nii.gz')
    cmd = 'fslnvols %s'%SUB
    cmd_split = cmd.split(' ')
    subprocess.run(cmd_split)

106
106


# Warm up 10/18/22
Based on what you know and information below. Get the TR value for the fmri_brain.nii.gz scan using subprocess. You will want to use the check_output function in subprocess. This will allow you to assign the output to a variable in python. 
```
X = subprocess.check_output(['ls','-a','<path to direcrory>'] 
```
The above would assign all the output from the ls command using the -a flag to the variable X
Check_output returns bytes (not STRINGS). So we also need to decode the byte into a string
```
Y = X.decode('utf-8')
```
Now Y is a string. 

You final TR should be a string with only the value of the TR, no extra characters

 You will use the following command:
 ```
 fslval <functional scan> pixdim4
 ```

In [5]:
import os
import subprocess
import pandas as pd
import glob

In [6]:
func_path = os.path.join(basepath, 'fmri1', 'fluency_task','fmri_brain.nii.gz')

In [7]:
os.path.exists(func_path)

True

In [8]:
# fslval func_path pixdim4 <- original command
# option 1
cmd = 'fslval %s pixdim4'%func_path
print(cmd)
cmd_split = cmd.split(' ')
print(cmd_split)
trs_alt = subprocess.check_output(cmd_split)
print(trs_alt)
trs_alt = trs_alt.decode('utf-8')
print(trs_alt)
final_trs = trs_alt.split(' ')[0]
print(final_trs)
# option 2
trs = subprocess.check_output(['fslval','%s'%(func_path),'pixdim4'])
# change the byte to string
trs = trs.decode('utf-8')
# get the number only
trs = trs.split(' ')[0]
print(trs)

fslval /Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri_brain.nii.gz pixdim4
['fslval', '/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri_brain.nii.gz', 'pixdim4']
b'4.200000 \n'
4.200000 

4.200000
4.200000


## Notes to self
* naming convention = sub-001
* subnum == the subject number
* OUTPUT = path to the output file = os.path.join(basepath,'level1','sub-%s'%subnum)
* TR = 4.2
* nvols = 106
* INPUT = path to the functional data

# Dictionary

In [9]:
inpt = os.path.join(basepath, 'fmri1','fluency_task','fmri_brain.nii.gz')

In [10]:
os.path.exists(inpt)

True

In [11]:
repl_dict = {}

In [12]:
repl_dict.update({'INPUT':inpt})

In [13]:
repl_dict

{'INPUT': '/Users/gracer/Desktop/fsl_course_data/fmri1/fluency_task/fmri_brain.nii.gz'}

In [27]:
outpath = os.path.join(basepath,'derivatives','sub-001.feat')

# For next time 10/25/22
## To do
Update repl_dict with all of your variables that in skeleton.fsf

In [37]:
datapath = os.path.join(basepath, 'fmri1','raw_data')

In [16]:
os.path.exists(datapath)

True

In [20]:
glob.glob(os.path.join(datapath,'sub-*'))

['/Users/gracer/Desktop/fsl_course_data/fmri1/raw_data/sub-002',
 '/Users/gracer/Desktop/fsl_course_data/fmri1/raw_data/sub-001']

In [22]:
outputpath = os.path.join(basepath,'fmri1','level1')

In [24]:
os.path.exists(outputpath)

False

naming convention = sub-001
subnum == the subject number
OUTPUT = path to the output file = os.path.join(basepath,'level1','sub-%s'%subnum)
TR = 4.2
nvols = 106
INPUT = path to the functional data

In [ ]:
example_dict = {'OUTPUT': 'actual path'}

In [49]:
example_list = ['x','y','z']
example_list[-1]
example_list.append('a')

In [38]:
#repl_dict = {'INPUT':[]}
datapath = os.path.join(basepath, 'fmri1','raw_data')
for x in glob.glob(os.path.join(datapath,'sub-*')):
    repl_dict = {} # initializing empty dictionary
    sub = x.split('/')[-1]
    inputpath = os.path.join(x,'fmri.nii.gz')
    if os.path.exists(inputpath) == False:
        print('no file')
    else:
        print('%s has fmri file'%sub)
    repl_dict.update({'INPUT':inputpath})
    OUT2 = os.path.join(basepath,'fmri1','level1')
    outputpath = os.path.join(OUT2,'%s.feat'%sub)
    repl_dict.update({'OUTPUT':outputpath})
    with open(os.path.join(basepath,'script','skeleton.fsf'),'r') as infile:
        tempfsf = infile.read()
        for key, value in repl_dict.items():
            print(key)
            tempfsf = tempfsf.replace(key,value)
        with open(os.path.join(basepath,'script','design_%s.fsf'%sub),'w') as outfile: #reason this wasn't working I forgot the w
            outfile.write(tempfsf)
        outfile.close()
    infile.close()
            
    
    
    

sub-002 has fmri file
INPUT
OUTPUT
sub-001 has fmri file
INPUT
OUTPUT


In [68]:
repl_dict

{'INPUT': '/Users/gracer/Desktop/fsl_course_data/fmri1/raw_data/sub-001/fmri.nii.gz',
 'OUTPUT': '/Users/gracer/Desktop/fsl_course_data/fmri1/level1/sub-001.feat'}

## For next time, annotate the loop the loop we made. Write what we are doing at each line

# Warm up

Write a loop that will populate a dictionary. The dictionary should be structured like this:  
key = feat folder name  
value = list of cope files in the stats folder (not the full path)   
Use the data in the fsl_course_data, in fmri2, in the paired_ttest folder, in the at folder

In [24]:
example_dict = {}
for featpath in glob.glob(os.path.join(basepath,'fmri2','paired_ttest','at','*.feat')):
    featkey = featpath.split('/')[-1]
    example_dict.update({featkey:[]})
    for copepath in glob.glob(os.path.join(featpath,'stats','cope*.nii.gz')):
        copevalue = copepath.split('/')[-1]
        example_dict[featkey].append(copevalue)
    
    

cope3.nii.gz
at_right.feat
cope1.nii.gz
at_right.feat
cope5.nii.gz
at_right.feat
cope2.nii.gz
at_right.feat
cope4.nii.gz
at_right.feat
cope6.nii.gz
at_right.feat
cope3.nii.gz
at_left.feat
cope1.nii.gz
at_left.feat
cope5.nii.gz
at_left.feat
cope2.nii.gz
at_left.feat
cope4.nii.gz
at_left.feat
cope6.nii.gz
at_left.feat


In [28]:
 example_dict['at_right.feat'][0:3]

['cope3.nii.gz', 'cope1.nii.gz', 'cope5.nii.gz']

In [39]:
for x in glob.glob(os.path.join(datapath,'sub-*')):
    repl_dict = {} # initializing empty dictionary
    sub = x.split('/')[-1]
    inputpath = os.path.join(x,'fmri.nii.gz')
    if os.path.exists(inputpath) == False:
        print('no file')
    else:
        print('%s has fmri file'%sub)
    repl_dict.update({'INPUT':inputpath})
    OUT2 = os.path.join(basepath,'fmri1','level1')
    outputpath = os.path.join(OUT2,'%s.feat'%sub)
    repl_dict.update({'OUTPUT':outputpath})
    # Find TR
    tr = findTR(inputpath)
    repl_dict.update({'TR':tr})
    with open(os.path.join(basepath,'script','skeleton.fsf'),'r') as infile:
        tempfsf = infile.read()
        for key, value in repl_dict.items():
            print(key)
            tempfsf = tempfsf.replace(key,value)
        with open(os.path.join(basepath,'script','design_%s.fsf'%sub),'w') as outfile: #reason this wasn't working I forgot the w
            outfile.write(tempfsf)
        outfile.close()
    infile.close()
            

sub-002 has fmri file
4.200000
INPUT
OUTPUT
TR
sub-001 has fmri file
4.200000
INPUT
OUTPUT
TR


In [31]:
# x can either be 'meow' or 'woof'
def whichanimal(x):
    if x == 'meow':
        return('this is a cat')
    elif x == 'woof':
        return('this is a dog')
    

In [29]:
def findTR(func_path):
    trs = subprocess.check_output(['fslval','%s'%(func_path),'pixdim4'])
    trs = trs.decode('utf-8')
    trs = trs.split(' ')[0]
    print(trs)
    return(trs)
    

<function __main__.findTR(func_path)>

In [ ]:
trs = subprocess.check_output(['fslval','%s'%(func_path),'pixdim4'])
# change the byte to string
trs = trs.decode('utf-8')
# get the number only
trs = trs.split(' ')[0]
print(trs)

# For next time 11/29
Update the loop with TR and nvols (see fslnvols) have it update the skeleton.fsf file with these values